skorch/callbacks.py

```diff
@@ -365,3 +365,79 @@ def on_epoch_end(self, net, **kwargs):
         self._sink(tabulated.rsplit('\n', 1)[-1], verbose)
         if self.sink is print:
             sys.stdout.flush()
+
+
+class Checkpoint(Callback):
+    """Save the model during training if the given metric improved.
+
+    This callback works by default in conjunction with the validation
+    scoring callback since it creates a `valid_loss_best` value
+    in the history which the callback uses to determine if this
+    epoch is save-worthy.
+
+    You can also specify your own metric to monitor or supply a
+    callback that dynamically evaluates whether the model should
+    be saved in this epoch.
+
+    Example:
+
+    >>> net = MyNet(callbacks=[Checkpoint()])
+    >>> net.fit(X, y)
+
+    Example using custom monitor:
+
+    >>> monitor = lambda net: all(net.history[-1, (
+    ...     'train_loss_best', 'valid_loss_best')]
+    >>> net = MyNet(callbacks=[Checkpoint(monitor=monitor)])
+    >>> net.fit(X, y)
+
+    Parameters
+    ----------
+    target : file-like object, str
+      File path to the file or file-like object.
+      See NeuralNet.save_params for details what this value may be.
+
+      If the value is a string you can also use format specifiers
+      to, for example, indicate the current epoch. Accessible format
+      values are `net`, `last_epoch` and `last_batch`.
+      Example to include last epoch number in file name:
+
+      >>> cb = Checkpoint(target="target_{last_epoch[epoch]}.pt")
+
+    monitor : str, function, None
+      Value of the history to monitor or callback that determines whether
+      this epoch should to a checkpoint. The callback takes the network
+      instance as parameter.
+
+      In case `monitor` is set to `None`, the callback will save
+      the network at every epoch.
+
+      **Note:** If you supply a lambda expression as monitor,
+      checkpointing will not work anymore as these values cannot
+      be pickled. You have to use importable functions instead.
```

Comment:
I read this note to mean that if a lambda is used (as in the example above), checkpointing would fail. This is, however, not true, as `save_params` only saves the model parameters. Pickling would not work, but that is True no matter what argument is set to be a lambda.
If something else is meant, please clarify.

Reply:
I missed that we only save the state dict, that's fine then.

You *can* pickle attributes that hold a reference to a function though, as long as it is importable.